Replicator timeout

MarkJr94 opened the pull request:
#### Problem

Replicator hangs if polling for leader fails or if unable to download ledger

#### Summary of Changes

Add an optional `timeout: Option<Duration>` to `Replicator::new`. Defaults to 10 seconds

Is currently only used when polling for the leader or downloading the ledger.

Ledger download times out if **0** entries are received within 10 seconds of initiation

Fixes #2379

sakridge commented on src/replicator.rs:
```diff
@@ -195,10 +218,25 @@ impl Replicator {
             sleep(Duration::from_millis(100));
+
+            let elapsed = start.elapsed();
+            received_so_far += entry_rx.try_recv().map(|v| v.len()).unwrap_or(0);
```
Seems like we will drop the receiver after this new(), will we leak entries into in the channel or will an error on the `send` occur?

MarkJr94 replied:
send will start producing errors once the receiver is dropped. I know that these `mpsc` channels have an unbounded buffer, but things only get pushed onto it if both ends are connected. I'm not sure how the sender on the other side will be affected by producing errors, I will look into that next. 

If this is a problem, is adding the receiver as a member of the `Replicator` struct a reasonable solution.

sakridge replied:
Yea, adding it to the Replicator struct should be fine.

sakridge commented on src/replicator.rs:
```diff
@@ -42,6 +43,9 @@ pub struct Replicator {
     pub retransmit_receiver: BlobReceiver,
+    // Currently unused. Stored to prevent the window service from experiencing errors when sending
+    #[allow(dead_code)]
+    entry_receiver: EntryReceiver,
```
So if we don't drop it, I don't we need to pull from it to avoid leaks? May need a separate thread to just pull from this receiver and discard the entries.

MarkJr94 replied:
This is true, the channel buffer could just keep growing unboundedly.

So one solution could be to remove the `?` after `entry_sender.send(consume_queue)?` and instead purposefully ignore that `Result`. But I'm not sure about whether that might cause issues. Should `recv_window` return an error if the `entry_sender.send()` returns an error?

sakridge commented:
@MarkJr94 heads up I merged the other PR, so can rebase on that now for this change.

MarkJr94 commented:
@sakridge  I rebased and implemented the ignoring of the results of the entry_sender, and nothing blew up in local tests. If the CI passes could you review this agau?
